ThreeBotPackages/tfgrid_solutions/tfgrid_solutions/chatflows/kubernetes_cluster_deploy.py

```diff
@@ -2,144 +2,182 @@
+                    masternodes.value + workernodes.value, self, **nodequery
+                )
+                self.master_nodes_selected = j.sal.reservation_chatflow.nodes_get(
+                    masternodes.value, farm_names=farms[: masternodes.value], **nodequery
+                )
+
+                self.worker_nodes_selected = j.sal.reservation_chatflow.nodes_get(
+                    workernodes.value, farm_names=farms[masternodes.value :], **nodequery
+                )
+                break
+            except StopChatFlow as e:
+                retry = True
+                self.md_show(e.msg)
+
+        self.user_form_data["Master number"] = masternodes.value
+        self.user_form_data["Workers number"] = workernodes.value
+        self.user_form_data["Cluster size"] = cluster_size_string.value
+
+    @j.baseclasses.chatflow_step(title="Access keys and secret")
+    def public_key_get(self):
+        self.user_form_data["SSH keys"] = self.upload_file(
+            """Please add your public ssh key, this will allow you to access the deployed container using ssh.
+                Just upload the ssh keys file with each key on a seperate line"""
         )
-        masternodes = form.int_ask(
-            "Please specify the number of master nodes", default=1, required=True, min=1, retry=retry
-        )  # minimum should be 1
-        workernodes = form.int_ask(
-            "Please specify the number of worker nodes", default=1, required=True, min=1, retry=retry
-        )  # minimum should be 1
-
-        form.ask()
-        cluster_size = sizes.index(cluster_size_string.value) + 1  # sizes are index 1
-        # Select nodes
-        if cluster_size == 1:
-            nodequery = {"sru": 50, "mru": 2, "cru": 1, "currency": currency}
-        else:
-            nodequery = {"sru": 100, "mru": 4, "cru": 2, "currency": currency}
-        try:
-            farms = j.sal.reservation_chatflow.farm_names_get(masternodes.value + workernodes.value, bot, **nodequery)
-            master_nodes_selected = j.sal.reservation_chatflow.nodes_get(
-                masternodes.value, farm_names=farms[: masternodes.value], **nodequery
+        self.ssh_keys_list = self.user_form_data["SSH keys"].split("\n")
+
+        self.user_form_data["Cluster secret"] = self.string_ask("Please add the cluster secret", default="secret")
+
+    @j.baseclasses.chatflow_step(title="Expiration")
+    def expiration_datetime(self):
+        self.expiration = self.datetime_picker(
+            "Please enter solution expiration time.",
+            required=True,
+            min_time=[3600, "Date/time should be at least 1 hour from now"],
+            default=j.data.time.epoch + 3900,
+        )
+        self.user_form_data["Solution expiration"] = j.data.time.secondsToHRDelta(self.expiration - j.data.time.epoch)
+
+    @j.baseclasses.chatflow_step(title="IP selection", disable_previous=True)
+    def ip_selection(self):
+
+        ipaddresses = list()
```

Comment:
why can't he go back from here? you would probably need to somehow be capable of removing the added nodes from the networks when back is selected but no permenanent changes would have been done yet

Reply:
https://github.com/threefoldtech/jumpscaleX_threebot/commit/7acbef7d8030b79316639179284a39c272825abe#diff-a5383166a7a1c1b6bd81d1638e7492c9L102